crates/sui-bridge-indexer/src/eth_bridge_indexer.rs

```diff
@@ -175,87 +183,151 @@ impl Datasource<RawEthData> for EthFinalizedSyncDatasource {
+    starting_checkpoint: u64,
+    data_sender: DataSender<RawEthData>,
+    indexer_metrics: BridgeIndexerMetrics,
+    bridge_metrics: Arc<BridgeMetrics>,
+) {
+    let eth_contracts_to_watch = HashMap::from_iter(
+        addresses
+            .iter()
+            .map(|address| (*address, starting_checkpoint)),
+    );
+
+    let (_, mut eth_events_rx, _) = EthSyncer::new(client.clone(), eth_contracts_to_watch)
+        .run(bridge_metrics.clone())
+        .await
+        .expect("Failed to start eth syncer");
+
+    // forward received events to the data sender
+    while let Some((_, block, logs)) = eth_events_rx.recv().await {
+        let raw_logs: Vec<RawEthLog> = logs
+            .into_iter()
+            .map(|log| RawEthLog {
+                block_number: block,  // Handle block number conversion
+                tx_hash: log.tx_hash, // Assuming `event` contains `tx_hash`
+                log: log.log,
+            })
+            .collect();
+
+        process_logs(raw_logs, provider.clone(), data_sender.clone(), block)
+            .await
+            .expect("Failed to process logs");
+        indexer_metrics.latest_committed_eth_block.set(block as i64);
+    }
 
-                    cached_blocks.insert(log.block_number, block.clone());
-                    block
-                };
+    panic!("Eth syncer stopped unexpectedly");
+}
 
-                if first_block == 0 {
-                    first_block = log.block_number;
-                }
+async fn retrieve_and_process_log_range(
+    client: Arc<EthClient<MeteredEthHttpProvier>>,
+    provider: Arc<Provider<Http>>,
+    addresses: Vec<EthAddress>,
+    starting_checkpoint: u64,
+    target_checkpoint: u64,
+    data_sender: DataSender<RawEthData>,
+    indexer_metrics: BridgeIndexerMetrics,
+) -> Result<(), Error> {
+    let Ok(Ok(logs)) = retry_with_max_elapsed_time!(
+        client.get_raw_events_in_range(addresses.clone(), starting_checkpoint, target_checkpoint),
+        Duration::from_secs(30000)
+    ) else {
+        panic!("Unable to get logs from provider");
+    };
+
+    process_logs(
+        logs,
+        provider.clone(),
+        data_sender.clone(),
+        target_checkpoint,
+    )
+    .await?;
+
+    indexer_metrics
+        .last_synced_eth_block
+        .set(target_checkpoint as i64);
+
+    Ok::<_, Error>(())
+}
 
-                let Ok(Ok(Some(transaction))) = retry_with_max_elapsed_time!(
-                    provider.get_transaction(log.tx_hash),
-                    Duration::from_secs(30000)
-                ) else {
-                    panic!("Unable to get transaction from provider");
-                };
+async fn process_logs(
+    logs: Vec<RawEthLog>,
+    provider: Arc<Provider<Http>>,
+    data_sender: DataSender<RawEthData>,
+    target_checkpoint: u64,
+) -> Result<(), Error> {
+    let mut data = Vec::new();
+    let mut cached_blocks: HashMap<u64, Block<H256>> = HashMap::new();
+    for log in logs {
```

Comment:
the way we do query `get_block` one by one here is not efficient. Let's add a TODO to add some parallelism